### PR TITLE
String switch optimizations (multi-compare and cache function)

### PR DESCRIPTION
**Pull request by zoecarver**
_This PR is a WIP._ In the case of a switch statement where a single string value is compared against many constant string values, the compiler will call a library function to do the string lookup. This patch is very similar to #11936 and heavily influenced by it. 

Resolves [SR-9440](https://bugs.swift.org/browse/SR-9440).

TODO:
- [ ] Fix the current implementation.
- [ ] Replace the call to `_findStringSwitchCase` with a call to `_findStringSwitchCaseWithCache`.

I have gotten a bit stuck on some issues in this patch and would love some help. I have left comments in the code about where I am getting stuck, but the main problem is the error I get because of passing `callExpr` as the subject of a `SwitchStmt`.

### CODE REVIEW

File: benchmark/single-source/StringSwitch.swift [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import TestsUtils
+
+public let StringEnum = BenchmarkInfo(
```
Comment (gottesmm) [inline benchmark/single-source/StringSwitch.swift]: This should be in a separate PR so we can see the perf difference.

File: lib/SILGen/SILGenPattern.cpp [hunk 2]
```diff
@@ -131,7 +131,7 @@ static bool isDirectlyRefutablePattern(const Pattern *p) {
   case PatternKind::Typed:
   case PatternKind::Var:
     return isDirectlyRefutablePattern(p->getSemanticsProvidingPattern());
-  }  
+  }
```
Comment (gottesmm) [inline lib/SILGen/SILGenPattern.cpp]: Please do this in a separate PR thing. It distracts from the review.

File: lib/SILGen/SILGenPattern.cpp [hunk 3]
```diff
@@ -2708,6 +2708,76 @@ void SILGenFunction::emitSwitchStmt(SwitchStmt *S) {
     return;
   }
 
+  // Optimize string switch statements by creating a cache that is searched through.
+  // TODO: how should I do the following? I couldn't find a StringType variable or similar.
+  if (subjectTy->getString() == "String") {
```
Comment (gottesmm) [inline lib/SILGen/SILGenPattern.cpp]: Why should this be done in SILGen?

**Review by gottesmm — COMMENTED** [review]
This is not an official review or anything, but I have a few questions.

File: lib/SILGen/SILGenPattern.cpp (hunk 3 above)
Comment (zoecarver) [reply lib/SILGen/SILGenPattern.cpp]: I am not very familiar with the swift codebase. I tried a few things in a few places, and this seemed to be the best. Is there a better place for it?

File: lib/SILGen/SILGenPattern.cpp (hunk 2 above)
Comment (zoecarver) [reply lib/SILGen/SILGenPattern.cpp]: Sorry, I will remove it.

File: lib/SILGen/SILGenPattern.cpp (hunk 3 above)
Comment (eeckstein) [reply lib/SILGen/SILGenPattern.cpp]: There are 2 possibilities: in SILGen and as an optimization pass.
With an optimization pass you would also catch code, which does not stem from switch statements, e.g. a series of if-statements with string compares. Also, SILGen is already a pretty big component and everything we can move out of SILGen helps to keep the code more maintainable.
On the other hand,  in an optimization pass you would need to do more pattern matching to detect the code pattern. This is problematic because it's easy to get something wrong with the pattern matching and also, it's kind of fragile because if any detail changes (e.g. in the String implementation), the pattern matching may fail.

Comment (zoecarver) [reply lib/SILGen/SILGenPattern.cpp]: What do you recommend? Seems like the optimization pass is better. I would love some help getting started with the pattern matching, maybe you can point me to the right file/give an example?

File: lib/SILGen/SILGenPattern.cpp [hunk 4]
```diff
@@ -2708,6 +2708,76 @@ void SILGenFunction::emitSwitchStmt(SwitchStmt *S) {
...
+    // find the cache function
+    SmallVector<ValueDecl *, 1> results;
+    // TODO: replace the following with a cache function
+    getModule().getASTContext().lookupInSwiftModule("_findStringSwitchCase", results);
```
Comment (Azoy) [inline lib/SILGen/SILGenPattern.cpp]: You can put this function in `include/swift/AST/KnownDecls.def` to cache it.

File: lib/SILGen/SILGenPattern.cpp [hunk 5]
```diff
@@ -2708,6 +2708,76 @@ void SILGenFunction::emitSwitchStmt(SwitchStmt *S) {
     return;
   }
 
+  // Optimize string switch statements by creating a cache that is searched through.
+  // TODO: how should I do the following? I couldn't find a StringType variable or similar.
```
Comment (Azoy) [inline lib/SILGen/SILGenPattern.cpp]: `if (subjectTy == ctx.getStringDecl()->getDeclaredType()) {`

File: lib/SILGen/SILGenPattern.cpp (hunk 3 above)
Comment (eeckstein) [reply lib/SILGen/SILGenPattern.cpp]: I think your approach in SILGen is good for now.
Some SILGen expert, like @gottesmm, should review the code in detail.

Comment (gottesmm) [reply lib/SILGen/SILGenPattern.cpp]: @eeckstein I disagree. We are actually creating new statements in SILGen. That is something we shouldn't do. SILGen wants to treat the AST as if it is walking it in a readonly way. If we want to do this sort of thing, I think we would need to do it in the type checker or something.

If we are going to do this in SILGen the implementation needs to be completely different and will involve changes to switch emission. If we do a pass, I think it will be simpler and TBH with proper testing, I don't think it will be an issue since we should be able to make a -Onone test that pattern matches this directly.

Comment (gottesmm) [reply lib/SILGen/SILGenPattern.cpp]: Do a git grep in SILGen for "::create" you will see there are no cases currently in SILGen where we are doing this.

Comment (zoecarver) [reply lib/SILGen/SILGenPattern.cpp]: It would seem odd to me that it would go in the type checker (but I also don't know much about the swift codebase). @gottesmm could you point me in the direction of another place something similar is being done? Maybe it would be good to look at other optimizations and do a similar thing.

Comment (gottesmm) [reply lib/SILGen/SILGenPattern.cpp]: That was my point. It would break layering to create nodes here and not in the type checker, yet would be weird to have the type checker synthesize this sort of code. That suggests that rather than creating nodes here and just running the already written code one would have to modify the underlying switch emission to understand how to do this peephole.

In terms of the type checker synthesizing code, it does it a bunch. Especially with derived conformances, creating inits, etc.

Comment (zoecarver) [reply lib/SILGen/SILGenPattern.cpp]: > yet would be weird to have the type checker synthesize this sort of code. 
Does that mean I should do it elsewhere? Where should this code go? It seems like the current place is not good which leaves the type checker or SIL optimizer.

Comment (gottesmm) [reply lib/SILGen/SILGenPattern.cpp]: No, I am saying you have two choices here:

1. Change the switch emission code to just recognize the pattern. This would not involve creating new nodes. Instead it would be an analysis of the current nodes.
2. Do it in the SILOptimizer.

Comment (zoecarver) [reply lib/SILGen/SILGenPattern.cpp]: Alright. Thank you for clarifying :) What would you recommend? As I have said before, my knowledge of the swift codebase is limited. Which would be most straightforward and most similar to other optimizations? Thanks again for all the help.

Comment (eeckstein) [reply lib/SILGen/SILGenPattern.cpp]: @gottesmm ok, makes sense

Comment (zoecarver) [reply lib/SILGen/SILGenPattern.cpp]: I have started adding an optimization pass in SILCodeMotion. How should I create a call to a function (the caching function)? Is there a way to do that with `ApplyInst` or something similar?

**Comment by zoecarver** [comment]
I re-wrote my patch inside of the SILOptimizer. I have left a few more comments about what I am doing and what I still need to do. This patch is a WIP but, I would like to know if I am on the right track. 

Currently, my biggest problem is `lookUpFunction` returns a `nullptr`. Any ideas about how to fix that? Maybe I need to look up the function globally? 

@eeckstein and @gottesmm, what do you think?

**Comment by gottesmm** [comment]
@zoecarver I would do this in SILCombine. Not in SILCodeMotion (I actually wrote this earlier before you posted the new patches, but my internet crapped out before it was submitted =/.

**Comment by zoecarver** [comment]
@gottesmm I'll put it there :)

File: lib/SILOptimizer/Transforms/SILCodeMotion.cpp [hunk 6]
```diff
@@ -32,6 +32,8 @@
 #include "llvm/Support/CommandLine.h"
 #include "llvm/Support/Debug.h"
 
+#include <iostream>
```
Comment (Azoy) [inline lib/SILOptimizer/Transforms/SILCodeMotion.cpp]: Please remove this and the print statement.

File: lib/SILOptimizer/Transforms/SILCodeMotion.cpp [hunk 7]
```diff
@@ -1759,6 +1787,36 @@ static bool processFunction(SILFunction *F, AliasAnalysis *AA,
...
+      SILFunction *fn = inst->getInitiallyReferencedFunction();
+      if (fn && !fn->empty())
+    		findStrCaseVals.push_back(fn->getArguments().front());
+    }
+    
+    // TODO: turn the findStrCaseVals into an array
+    
+    // Find the function decl
+    SmallVector<ValueDecl *, 1> results;
+    Builder.getASTContext().lookupInSwiftModule("_findStringSwitchCase", results);
```
Comment (Azoy) [inline lib/SILOptimizer/Transforms/SILCodeMotion.cpp]: Like I said earlier, you can put this function in KnownDecls.def that way it can be cached.

File: lib/SILOptimizer/Transforms/SILCodeMotion.cpp (hunk 6 above)
Comment (gottesmm) [reply lib/SILOptimizer/Transforms/SILCodeMotion.cpp]: @zoecarver  More meta, in general we do not use iostream in LLVM based projects. I would familiarize oneself with these two documents: 

1. https://www.llvm.org/docs/ProgrammersManual.html.
2. https://www.llvm.org/docs/CodingStandards.html

For instance the iostream issue is here: https://www.llvm.org/docs/CodingStandards.html#include-iostream-is-forbidden.

Comment (gottesmm) [reply lib/SILOptimizer/Transforms/SILCodeMotion.cpp]: @zoecarver  More meta, in general we do not use iostream in LLVM based projects. I would familiarize oneself with these two documents: 

1. https://www.llvm.org/docs/ProgrammersManual.html.
2. https://www.llvm.org/docs/CodingStandards.html

For instance the iostream issue is here: https://www.llvm.org/docs/CodingStandards.html#include-iostream-is-forbidden.

Comment (zoecarver) [reply lib/SILOptimizer/Transforms/SILCodeMotion.cpp]: Thanks for the links. Forgot to remove it before pushing my code. As I said, this is very much a work in progress.

File: lib/SILOptimizer/Transforms/SILCodeMotion.cpp (hunk 7 above)
Comment (zoecarver) [reply lib/SILOptimizer/Transforms/SILCodeMotion.cpp]: Will do. Will that fix the `lookUpFunction ` issue?

File: lib/SILOptimizer/Transforms/SILCodeMotion.cpp (hunk 6 above)
Comment (gottesmm) [reply lib/SILOptimizer/Transforms/SILCodeMotion.cpp]: @zoecarver No worries! = ).

Comment (gottesmm) [reply lib/SILOptimizer/Transforms/SILCodeMotion.cpp]: The only difference with the coding standards is that we generally make functions and variables camelCase, not CamelCase. (As always, match the style of the code around it, but if you are writing new code, new functions, plz do it this way = ).

Comment (zoecarver) [reply lib/SILOptimizer/Transforms/SILCodeMotion.cpp]: I will try to do that. I have seen a lot of TitleCase variables. Should variables be TitleCase or camelCase?

Comment (gottesmm) [reply lib/SILOptimizer/Transforms/SILCodeMotion.cpp]: camelCase. We have been doing the change slowly over time in new code.

File: lib/SILOptimizer/Transforms/SILCodeMotion.cpp (hunk 7 above)
Comment (Azoy) [reply lib/SILOptimizer/Transforms/SILCodeMotion.cpp]: What you probably want to do is call something like `emitApplyOfLibraryIntrinsic` once you get the func decl from the astcontext.

Comment (zoecarver) [reply lib/SILOptimizer/Transforms/SILCodeMotion.cpp]: Thanks, @Azoy, that is super helpful. Looks like `emitApplyOfLibraryIntrinsic` is a member of `SILGenFunction`. How can I get access to that in the SILOptimizer?

Comment (Azoy) [reply lib/SILOptimizer/Transforms/SILCodeMotion.cpp]: Oh my apologies. You’re correct that this is only available in SILGen (whoops). What you can do however is something like the following:
```cpp
auto mangled = SILDeclRef(findFn, SILDeclRef::Kind::Func).mangle();
auto *silFn = F->getModule().findFunction(mangled, SILLinkage::PublicExternal);
```

Comment (zoecarver) [reply lib/SILOptimizer/Transforms/SILCodeMotion.cpp]: Thanks, I really appreciate your help. Unfortunately, that still gave me a `nullptr`. Any other ideas? I've looked through some other places `findFunction` is used in the SILOptimizer and it seems like it _should_ work but, it doesn't for some reason. Maybe we can hardcode the mangled function (like what's done for `_swift_isClassOrObjCExistentialType`)?

**Comment by zoecarver** [comment]
Re-implemented #28260.
